src/com/can/CANLayer.cpp

```diff
@@ -0,0 +1,157 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Malte Grave, LIT CPS Johannes Kepler Universität,
+ *                                 github.com/gravemalte
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Malte Grave - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+
+#include <cstdint>
+#include <cstring>
+#include <fcntl.h>
+#include <iostream>
+#include <linux/can.h>
+#include <string>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+
+#include "CANLayer.h"
+#include "basecommfb.h"
+#include "comtypes.h"
+#include "datatype.h"
+#include "devlog.h"
+#include "forte_any.h"
+#include "forte_byte.h"
+
+using namespace forte::com_infra;
+
+CCANComLayer::CCANComLayer(CComLayer *paUpperLayer, CBaseCommFB *paComFB) :
+	CComLayer(paUpperLayer, paComFB) {}
+
+EComResponse CCANComLayer::openConnection(char *paLayerParameter) {
+	EComResponse eRetVal = e_InitInvalidId;
+	CParameterParser parser(paLayerParameter, ';', eCANComParamterAmount);
+	if (eCANComParamterAmount != parser.parseParameters()) {
+		return eRetVal;
+	}
+
+	setupCANInternals(parser);
+
+	switch (mFb->getComServiceType()) {
+	case e_Server:
+		DEVLOG_DEBUG("[CAN Layer] the layer only supports the PUB/SUB "
+					 "pattern. Please use a Publisher block to send data.");
+		eRetVal = e_InitTerminated;
+		break;
+	case e_Client:
+		DEVLOG_DEBUG("[CAN Layer] the layer only supports the PUB/SUB "
+					 "pattern. Please use a Subscribe block to receive data.");
+		eRetVal = e_InitTerminated;
+		break;
+	case e_Publisher:
+		// is handled via sendData
+		break;
+	case e_Subscriber:
+		// is handled via recvData
+		break;
+	}
+
+	DEVLOG_DEBUG("[CAN Layer] Opening connection\n");
+
+	CFDSelectHandler::TFileDescriptor fd = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+
+	if (CFDSelectHandler::scmInvalidFileDescriptor == fd) {
+		DEVLOG_DEBUG("[CAN Layer] Socket can't be created.\n");
+		return forte::com_infra::e_InitInvalidId;
+	}
+
+	mCANSocket = fd;
+	std::strcpy(mIfr.ifr_name, mCANParam.interfaceName.c_str());
```

Comment:
you need to make sure that the given interface name is shorter or equal to `IFNAMSIZ - 1`